File: backend/app/services/receive_batch_service.py
import uuid
from datetime import date, datetime

from app.extensions import db
from app.models.package import Package
from app.models.receive_batch import ReceiveBatch
from app.models.user import User


def _next_batch_code(receive_date: date) -> str:
    prefix = receive_date.strftime("%m%d")
    pattern = f"RB-{prefix}-%"
    count = (
        ReceiveBatch.query.filter(ReceiveBatch.batch_code.like(pattern)).count()
    )
    return f"RB-{prefix}-{count + 1:02d}"


def create_receive_batch(
    *,
    reference: str | None = None,
    receive_date: date | None = None,
    note: str | None = None,
    created_by: User | None = None,
) -> ReceiveBatch:
    receive_date = receive_date or date.today()
    batch_code = _next_batch_code(receive_date)
    label = (reference or "").strip() or batch_code

    batch = ReceiveBatch(
        batch_code=batch_code,
        reference=label,
        receive_date=receive_date,
        status="open",
        note=note.strip() if note else None,
        created_by_id=created_by.id if created_by else None,
    )
    db.session.add(batch)
    db.session.commit()
    return batch


def list_receive_batches(
    *,
    status: str | None = None,
    limit: int = 50,
    offset: int = 0,
) -> tuple[list[ReceiveBatch], int]:
    query = ReceiveBatch.query
    if status:
        query = query.filter(ReceiveBatch.status == status)
    query = query.order_by(ReceiveBatch.receive_date.desc(), ReceiveBatch.created_at.desc())
    total = query.count()
    batches = query.offset(offset).limit(limit).all()
    return batches, total


def get_receive_batch(batch_id: uuid.UUID) -> ReceiveBatch | None:
    return db.session.get(ReceiveBatch, batch_id)


def assert_open_receive_batch(batch: ReceiveBatch) -> None:
    if batch.status != "open":
        raise ValueError(f"Receive batch {batch.batch_code} is closed")


def assign_package_to_receive_batch(package: Package, batch: ReceiveBatch) -> None:
    assert_open_receive_batch(batch)
    package.receive_batch_id = batch.id
    package.updated_at = datetime.utcnow()


def resolve_receive_batch_id(raw_id: str | None) -> ReceiveBatch | None:
    if not raw_id:
        return None
    try:
        batch_id = uuid.UUID(str(raw_id))
    except (TypeError, ValueError):
        raise ValueError("Invalid receive_batch_id")
    batch = get_receive_batch(batch_id)
    if not batch:
        raise ValueError("Receive batch not found")
    assert_open_receive_batch(batch)
    return batch

File: server/src/services/receiveBatchService.ts
import { and, desc, eq, like } from 'drizzle-orm'

import { db } from '../db/index.js'
import { packages, receiveBatches } from '../db/schema/index.js'
import type { packages as packagesTable, receiveBatches as receiveBatchesTable } from '../db/schema/index.js'
import type { UserRow } from '../db/schema/index.js'
import { receiveBatchToDict } from '../lib/serializers/receiveBatch.js'

type PackageRow = typeof packagesTable.$inferSelect
type ReceiveBatchRow = typeof receiveBatchesTable.$inferSelect

function nextBatchCode(receiveDate: string): string {
  const prefix = receiveDate.slice(5, 7) + receiveDate.slice(8, 10)
  return `RB-${prefix}-01`
}

export async function createReceiveBatch(options: {
  reference?: string | null
  receiveDate?: string | null
  note?: string | null
  createdBy?: UserRow | null
}): Promise<ReceiveBatchRow> {
  const receiveDate = options.receiveDate ?? new Date().toISOString().slice(0, 10)
  const pattern = `RB-${receiveDate.slice(5, 7)}${receiveDate.slice(8, 10)}-%`
  const existing = await db
    .select()
    .from(receiveBatches)
    .where(like(receiveBatches.batchCode, pattern))

  const batchCode =
    existing.length > 0
      ? `RB-${receiveDate.slice(5, 7)}${receiveDate.slice(8, 10)}-${String(existing.length + 1).padStart(2, '0')}`
      : nextBatchCode(receiveDate)

  const label = (options.reference ?? '').trim() || batchCode

  const now = new Date()
  const [batch] = await db
    .insert(receiveBatches)
    .values({
      id: crypto.randomUUID(),
      batchCode,
      reference: label,
      receiveDate,
      status: 'open',
      note: options.note?.trim() || null,
      createdById: options.createdBy?.id ?? null,
      createdAt: now,
      updatedAt: now,
    })
    .returning()

  return batch
}

export async function listReceiveBatches(options?: {
  status?: string | null
  limit?: number
  offset?: number
}): Promise<[ReceiveBatchRow[], number]> {
  const limit = options?.limit ?? 50
  const offset = options?.offset ?? 0

  let rows = await db.select().from(receiveBatches).orderBy(desc(receiveBatches.receiveDate), desc(receiveBatches.createdAt))

  if (options?.status) {
    rows = rows.filter((b) => b.status === options.status)
  }

  const total = rows.length
  return [rows.slice(offset, offset + limit), total]
}

export async function getReceiveBatch(batchId: string): Promise<ReceiveBatchRow | null> {
  const [batch] = await db.select().from(receiveBatches).where(eq(receiveBatches.id, batchId)).limit(1)
  return batch ?? null
}

export function assertOpenReceiveBatch(batch: ReceiveBatchRow): void {
  if (batch.status !== 'open') {
    throw new Error(`Receive batch ${batch.batchCode} is closed`)
  }
}

export function assignPackageToReceiveBatch(pkg: PackageRow, batch: ReceiveBatchRow): void {
  assertOpenReceiveBatch(batch)
  ;(pkg as { receiveBatchId?: string | null }).receiveBatchId = batch.id
  ;(pkg as { updatedAt?: Date }).updatedAt = new Date()
}

export async function resolveReceiveBatchId(rawId: string | null | undefined): Promise<ReceiveBatchRow | null> {
  if (!rawId) return null
  const batch = await getReceiveBatch(rawId)
  if (!batch) throw new Error('Receive batch not found')
  assertOpenReceiveBatch(batch)
  return batch
}

export { receiveBatchToDict }
